Build-process.md

```diff
@@ -0,0 +1,116 @@
+## Table of contents
+
+* [Introduction](#introduction)
+* [Build script](#build-script)
+* [Build modes](#build-modes)
+  * [Build mode constants](#build-mode-constants)
+  * [Dev mode](#dev-mode)
+  * [Prod mode](#prod-mode)
+    * [Plain prod mode](#plain-prod-mode)
+    * [Minify only third-party libraries](#minify-only-third-party-libraries)
+    * [Maintenance mode](#maintenance-mode)
+    * [Deploy mode](#deploy-mode)
+    * [Unused: Deparallelized terser mode](#unused-deparallelized-terser-mode)
+* [Source maps](#source-maps)
+
+## Introduction
+
+Before we can run the Oppia server, we have to build it. This build process transforms the files defining our application (e.g. code, images, configuration files, etc.) from the developer-friendly format we store in GitHub into a form that can be executed to run the server. If you are running the local development server, this build process makes minimal changes, but when deploying to production, we do much more optimization.
+
+This page documents how the build process works in each of its modes. For most developers the build steps are handled automatically and can be ignored; however, some bugs appear only in certain build modes. For example, a bug might only happen in prod mode, in which case a developer will be unable to reproduce it in dev mode.
+
+## Build script
+
+The build process is performed by [`scripts/build.py`](https://github.com/oppia/oppia/blob/develop/scripts/build.py). If you ever want to know exactly how a particular part of the build process works, you should check out the code there.
+
+## Build modes
+
+There are two primary build modes: dev mode and prod mode. When running in prod mode, you can also enable maintenance mode or deploy mode, or you can choose to only minify third-party libraries. This means the build mode hierarchy looks like this:
+
+* Dev mode
+* Prod mode
+  * Maintenance mode
+  * Deploy mode
```

Comment:
Just a note: both maintenance mode and prod mode can be enabled together, we usually expect the maintenance mode be enabled with deploy mode so it might be possible to include the maintenance mode under deploy mode.

Reply:
I added a note that maintenance and deploy mode can be enabled together, but I think it still makes sense to discuss them separately because maintenance mode can also be enabled locally (e.g. for testing maintenance mode)